Inferences/ForwardDemodulation.cpp

```diff
@@ -169,29 +175,48 @@ bool ForwardDemodulationImpl<combinatorySupSupport>::perform(Clause* cl, Clause*
           continue;
         }
 
+        // encompassing demodulation is fine when rewriting the smaller guy
+        if (toplevelCheck && _encompassing) { 
+          // this will only run at most once; 
+          // could have been factored out of the getGeneralizations loop, 
+          // but then it would run exactly once there
+          Ordering::Result litOrder = ordering.getEqualityArgumentOrder(lit);
+          if ((trm==*lit->nthArgument(0) && litOrder == Ordering::LESS) || 
+              (trm==*lit->nthArgument(1) && litOrder == Ordering::GREATER)) {
+            toplevelCheck = false;
+          }
+        }
+
         if(toplevelCheck) {
           TermList other=EqHelper::getOtherEqualitySide(lit, trm);
           Ordering::Result tord=ordering.compare(rhsS, other);
           if(tord!=Ordering::LESS && tord!=Ordering::LESS_EQ) {
-            Literal* eqLitS=qr.substitution->applyToBoundResult(qr.literal);
-            bool isMax=true;
-            for(unsigned li2=0;li2<cLen;li2++) {
-              if(li==li2) {
-          continue;
+            if (_encompassing) {
+              // last chance, if the matcher is not a renaming
+              if (qr.substitution->isRenamingOn(qr.term,true /* we talk of result term */)) {
+                continue; // under _encompassing, we know there are no other literals in cl
```

Comment:
Implementation doesn't seem to capture full generality of rule. Substitution can be a renaming on the lareger side, so long as `r\theta < t` (condition (v) in the paper)

Reply:
Agreed